p2p/handshake/handshake.go

```diff
@@ -0,0 +1,278 @@
+	return func(w *transportWrapper) {
+		w.attempts = n
+	}
+}
+
+type transportWrapper struct {
+	transport.Transport
+	nc            NetworkCookie
+	logger        log.Log
+	timeout       time.Duration
+	retryInterval time.Duration
+	attempts      int
+}
+
+var (
+	_ transport.Transport = &transportWrapper{}
+	_ io.Closer           = &transportWrapper{}
+)
+
+// MaybeWrapTransport adds a handshake wrapper around the Transport if
+// network cookie nc is not empty, otherwise it just returns the
+// transport.
+func MaybeWrapTransport(t transport.Transport, nc NetworkCookie, opts ...Option) transport.Transport {
+	if nc.Empty() {
+		return t
+	}
+	tr := &transportWrapper{
+		Transport:     t,
+		nc:            nc,
+		logger:        log.NewNop(),
+		timeout:       handshakeTimeout,
+		retryInterval: handshakeRetryInterval,
+		attempts:      handshakeAttempts,
+	}
+	for _, opt := range opts {
+		opt(tr)
+	}
+	return tr
+}
+
+func (tr *transportWrapper) Dial(ctx context.Context, raddr ma.Multiaddr, p peer.ID) (transport.CapableConn, error) {
+	c, err := tr.Transport.Dial(ctx, raddr, p)
+	if err != nil {
+		return nil, err
+	}
+
+	for i := 0; i < tr.attempts; i++ {
+		var retry bool
+		retry, err = tr.handshake(ctx, c)
+		switch {
+		case err == nil:
+			return c, nil
+		case retry:
+			tr.logger.With().Error("transport wrapper handshake error",
+				log.Int("attempt", i+1),
+				log.Int("handshakeAttempts", tr.attempts),
+				log.Err(err))
+			select {
+			case <-ctx.Done():
+				return nil, ctx.Err()
+			case <-time.After(handshakeRetryInterval):
+			}
+		default:
+			return nil, err
+		}
+	}
+
+	return nil, err
+}
+
+func (tr *transportWrapper) handshake(ctx context.Context, c transport.CapableConn) (retry bool, err error) {
+	attemptCtx, cancel := context.WithTimeout(ctx, handshakeTimeout)
+	defer cancel()
+	s, err := c.OpenStream(attemptCtx)
+	if err != nil {
+		return true, err
+	}
+
+	defer s.Close()
+	s.SetDeadline(time.Now().Add(tr.timeout))
+
+	if err := writeCookie(s, tr.nc); err != nil {
+		return true, err
+	}
+
+	remoteCookie, mayHaveCookie, err := readCookie(s)
+	if err != nil {
+		return mayHaveCookie, err
+	}
+
+	if !tr.nc.Equal(remoteCookie) {
+		return false, fmt.Errorf("cookie mismatch: %s instead of expected %s", remoteCookie, tr.nc)
+	}
+
+	return false, nil
+}
+
+func (tr *transportWrapper) Listen(laddr ma.Multiaddr) (transport.Listener, error) {
+	l, err := tr.Transport.Listen(laddr)
+	if err != nil {
+		return nil, ma.ErrProtocolNotFound
+	}
+	return &listenerWrapper{
+		Listener: l,
+		nc:       tr.nc,
+		logger:   tr.logger,
+		timeout:  tr.timeout,
+		attempts: tr.attempts,
+	}, nil
+}
+
+func (tr *transportWrapper) Close() error {
+	if c, ok := tr.Transport.(io.Closer); ok {
+		return c.Close()
+	}
+
+	return nil
+}
+
+type listenerWrapper struct {
+	transport.Listener
+	nc       NetworkCookie
+	logger   log.Log
+	timeout  time.Duration
+	attempts int
+}
+
+var _ transport.Listener = &listenerWrapper{}
```

Comment:
same as above, this seems redundant when `listenWrapper` already is a `transport.Listener` via embedding

Reply:
Also, same as above: "overridden" methods can have wrong signature (e.g. after `go-libp2p` version update) and having these checks fail might save us a little bit of time when upgrading the deps someday in future